### ISSUE TITLE
Finer grained permission for writing commit status

### ISSUE THREAD

**Issue by ascandella** [go-gitea/gitea#32625]
### Feature Description

Currently, to generate a token that can create commit statuses via the API, it must have `write:repository`. This is less than ideal from a security perspective, where I'd like to give a token to some automation tools that can write commit statuses but not write to the repository.

An example of this is a deployment system. I use ArgoCD with my Gitea instance and I have the notifications controller set up to send commit statuses when applications are synced successfully or with an error. In this context I don't want to have a token anywhere that can write to my repository. 

Proposal: create a new permissions for commit statuses (read, and read+write) and update the API to only require commit statuses privileges to read/write commit status. Additionally, for backwards compatibility and sanity leave the `repository` permission as sufficient for commit statuses as well.

If this idea is acceptable, I'd be happy to implement it myself.

**Comment by wxiaoguang** [comment]
Do you mean something like

* `write:repository`: allow writing commit status (as before, for backwards compatibility) 
* `write:commit_status`: also allow writing commit status (but not others)

?

**Comment by ascandella** [comment]
> Do you mean something like
> 
> 
> 
> * `write:repository`: allow writing commit status (as before, for backwards compatibility) 
> 
> * `write:commit_status`: also allow writing commit status (but not others)
> 
> 
> 
> ?

Exactly. And probably a read:commit_status although I don't know of a great use for that without read:repository access (which would still be sufficient for reading commit statuses)... thoughts?

**Comment by wxiaoguang** [comment]
Yup, the "read" permission for a commit status seems quite strange, no idea whether it should really be introduced (while it is also strange not introducing it .....)

----

Overall I think your idea is pretty good (acceptable to me). The real challenge is that how to implement it (the token scope system is complex 🤣 )

**Comment by ascandella** [comment]
Ok cool, I'll start digging into the code in the next few weeks and see what I can come up with. Thanks!

**Comment by ascandella** [comment]
For reference GitHub as a read-only commit statuses option:

<img width="812" alt="image" src="https://github.com/user-attachments/assets/a69fae21-d313-47a7-beba-7e9aa3cc6669">

https://docs.github.com/en/rest/commits/statuses?apiVersion=2022-11-28#get-the-combined-status-for-a-specific-reference

**Comment by not7cd** [comment]
@ascandella, were you able to dig something up? I have a similar use-case to yours and I'm interested in this feature too.

**Comment by not7cd** [comment]
I have read through the code a bit to figure out where to start. This is the model, I guess we would need to add `AccessTokenScopeCategoryCommitStatus`
https://github.com/go-gitea/gitea/blob/c4fbccc4ec714a6575cf97da827fd9b27133094b/models/auth/access_token_scope.go#L17-L27

Then add this additional scope here?
https://github.com/go-gitea/gitea/blob/c4fbccc4ec714a6575cf97da827fd9b27133094b/routers/api/v1/api.go#L1404-L1408

via this middleware
https://github.com/go-gitea/gitea/blob/c4fbccc4ec714a6575cf97da827fd9b27133094b/routers/api/v1/api.go#L319

**Comment by ascandella** [comment]
@not7cd I started on it here: https://github.com/ascandella/gitea/tree/ascandella/commit-status-token-scope

I ended up getting tripped up and unable to unwind the permission model in my head as to how to actually plumb this through, so I put it on the backburner. If you want to pick it up my branch may be a helpful place to start (or I may be totally off base 😀 )

**Comment by not7cd** [comment]
Thanks for sharing your branch. It looks like a proper foundation for a final PR. Only few TODOs to resolve. 

I'm unsure if this covers all endpoints. It's a bit hard to navigate api routes.
I'm mainly interested in this endpoint, and it should be covered already:

https://github.com/jenkinsci/gitea-plugin/blob/dd9e9b975949fb44cd7ca9230439e0ef45227928/src/main/java/org/jenkinsci/plugin/gitea/client/impl/DefaultGiteaConnection.java#L617-L630

There is another related endpoint, I think this is convenience endpoint to read status from commit route?
https://github.com/go-gitea/gitea/blob/0739595dcd2f24291a9bd04474808e6ae262f9d6/routers/api/v1/api.go#L1404-L1413

@lunny, since this proposal was accepted, would you prefer to have an open PR and move discussion there? Or should we explore implementation avenues in this issue?

**Comment by lunny** [comment]
Please open a pull request so we can continue the discussion there.
